contracts/ConsensusUtils.sol

```diff
@@ -328,13 +388,54 @@ contract ConsensusUtils is EternalStorage, ValidatorSet {
     return false;
   }
 
+  function _jailValidator(address _address) internal {
+    _pendingValidatorsRemove(_address);
+    _addJailedValidator(_address);
+    _setJailRelease(_address);
+    _resetStrikeReset(_address);
+  }
+
+  function _maintenance(address _address) internal {
+    _pendingValidatorsRemove(_address);
+    _addJailedValidator(_address);
+  }
+
   function _setPendingValidatorsAtPosition(uint256 _p, address _address) internal {
     addressArrayStorage[PENDING_VALIDATORS][_p] = _address;
   }
 
+  function _setJailedValidatorsAtPosition(uint256 _p, address _address) internal {
+    addressArrayStorage[JAILED_VALIDATORS][_p] = _address;
+  }
+
   function _pendingValidatorsAdd(address _address) internal {
+    require(isJailed(_address) == false);
     addressArrayStorage[PENDING_VALIDATORS].push(_address);
-    _setValidatorFee(_address, DEFAULT_VALIDATOR_FEE);
+  }
+
+  function _addJailedValidator(address _address) internal {
+    addressArrayStorage[JAILED_VALIDATORS].push(_address);
+  }
+
+  function _jailedValidatorRemove(address _address) internal {
+    bool found = false;
+    uint256 removeIndex;
+    for (uint256 i; i < jailedValidatorsLength(); i++) {
+      if (_address == jailedValidatorsAtPosition(i)) {
+        removeIndex = i;
+        found = true;
```

Comment:
One optimization you can do, is to add a `break` here so when a validator is found we can leave the loop